### PR TITLE
reject registering users using habitica.com or habitrpg.com emails

### PR DESCRIPTION
**Pull request by librarianmage**
# Changes:

Rejects registering users that use a habitrpg.com or a habitica.com email account (fixing #7633)

---

UUID: 95f16006-1504-4a9f-95b6-0c7314dbd7ee

### CODE REVIEW

File: website/server/controllers/api-v3/auth.js [hunk 1]
```diff
@@ -75,6 +75,10 @@ api.registerLocal = {
       email: {
         notEmpty: {errorMessage: res.t('missingEmail')},
         isEmail: {errorMessage: res.t('notAnEmail')},
+        matches: {
+          options: ['.+@(?!(?:(?:habitrpg\.com$)|(?:habitica\.com$))).+', 'i'],
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/auth.js]: This is a messy regex. Would prefer a custom validate method.

Comment (librarianmage) [inline website/server/controllers/api-v3/auth.js]: Not sure if that can be done with the schema, but I can look.

Comment (crookedneighbor) [inline website/server/controllers/api-v3/auth.js]: http://mongoosejs.com/docs/validation.html

**Comment by librarianmage** [comment]
@crookedneighbor Can you help me find why all of the integration tests broke?

**Comment by librarianmage** [comment]
@crookedneighbor if it helps, I've reasonably determined that it is occurring in the generateUser function

File: website/server/controllers/api-v3/auth.js [hunk 2]
```diff
@@ -75,6 +75,10 @@ api.registerLocal = {
       email: {
         notEmpty: {errorMessage: res.t('missingEmail')},
         isEmail: {errorMessage: res.t('notAnEmail')},
+        notEmailDomains: {
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/auth.js]: This is why all the tests are failing.

Comment (crookedneighbor) [inline website/server/controllers/api-v3/auth.js]: Why did you move the validation here instead of using a validate function on the model like I asked for?

Comment (librarianmage) [inline website/server/controllers/api-v3/auth.js]: @crookedneighbor I guess I misunderstood. I'm sorry.

Comment (librarianmage) [inline website/server/controllers/api-v3/auth.js]: Do you want it in the `middlewares` section?

Comment (crookedneighbor) [inline website/server/controllers/api-v3/auth.js]: Take a look at how it's done here: https://github.com/HabitRPG/habitrpg/pull/7655/files#diff-1f63728ba8da6291c25b8acf11aab121R32

Comment (librarianmage) [inline website/server/controllers/api-v3/auth.js]: Thanks!

File: website/server/models/user/schema.js [hunk 3]
```diff
@@ -24,7 +24,21 @@ let schema = new Schema({
     local: {
       email: {
         type: String,
-        validate: [validator.isEmail, shared.i18n.t('invalidEmail')],
+        validate: [{
+          validator: validator.isEmail,
+          msg: shared.i18n.t('invalidEmail'),
+        }, {
+          validator: function notEmailDomains (email) {
+            let domains = ['habitica\\.com', 'habitrpg\\.com'];
```
Comment (crookedneighbor) [inline website/server/models/user/schema.js]: I'd put these as a const near the top of the file:

``` js
const INVALID_DOMAINS = ['habitica.com', 'habitrpg.com'];
```

And then in the validator, I'd just do this:

``` js
validator (email) {
  return INVALID_DOMAINS.every(domain => email.indexOf(domain) === -1);
}
```

Comment (librarianmage) [inline website/server/models/user/schema.js]: Okay, I just didn't want to exclude an email like `habitica.com@example.com` (even though that is a really contrived example)

Comment (crookedneighbor) [inline website/server/models/user/schema.js]: Good point, I had intended to make it:

``` js
return INVALID_DOMAINS.every(domain => email.indexOf(`@${domain}`) === -1);
```

And just forgot. That should take care of that edge case.

Comment (librarianmage) [inline website/server/models/user/schema.js]: that might work. Would you still want an email like `example@test.habitica.com` to fail?

Comment (librarianmage) [inline website/server/models/user/schema.js]: If you do, would

``` javascript
return INVALID_DOMAINS.every(domain => new RegExp(`${domain}$`, 'i').test(email) === false);
```

work?

Comment (librarianmage) [inline website/server/models/user/schema.js]: that would also fix failing an email like `not.fail@habitica.com.au`

File: website/server/models/user/schema.js [hunk 4]
```diff
@@ -9,6 +9,8 @@ import {
 
 const Schema = mongoose.Schema;
 
+const INVALID_DOMAINS = ['habitica.com', 'habitrpg.com'];
```
Comment (awitherow) [inline website/server/models/user/schema.js]: Do you think there will be more invalid domains eventually? If the list here grows, perhaps exporting stuff to a config file of sorts would be useful. Perhaps also over-engineering 😉

Comment (librarianmage) [inline website/server/models/user/schema.js]: Perhaps, if Habitica goes through another rebranding 😉

Comment (crookedneighbor) [inline website/server/models/user/schema.js]: Full disclosure: I was going to convert this to an env variable while merging.

Comment (librarianmage) [inline website/server/models/user/schema.js]: Makes sense

File: website/server/models/user/schema.js [hunk 5]
```diff
@@ -24,7 +26,15 @@ let schema = new Schema({
     local: {
       email: {
         type: String,
-        validate: [validator.isEmail, shared.i18n.t('invalidEmail')],
+        validate: [{
+          validator: validator.isEmail,
+          msg: shared.i18n.t('invalidEmail'),
+        }, {
+          validator: function notEmailDomains (email) {
```
Comment (crookedneighbor) [inline website/server/models/user/schema.js]: Let's use the object shorthand here:

``` js
validator (email) {
```

**Comment by librarianmage** [comment]
Finished!

**Comment by librarianmage** [comment]
:bell: This should be finished.

**Comment by crookedneighbor** [comment]
I elected to remove the client side portions of this after fixing a bug that prevented the custom messages from being displayed on the client. https://github.com/HabitRPG/habitrpg/commit/0150b355cb25067b23969b13b1048756548f8927

I've notated this toward your next tier.
